R/doubletFinder.R

```diff
@@ -87,13 +97,151 @@ doubletFinder <- function(seu, expected.doublets = 0, proportion.artificial = 0.
+#' whether the online PCA method should be used. Online PCA should only be used for very large
+#' datasets.
+#' @return (loom) Nothing; loom objects are modified in place. This function will add pANN values
+#' and doublet predictions to column attributes.
+#' @describeIn doubletFinder Doublet detection in single-cell RNA sequencing data
+#' @export doubletFinder.loom
+#' @method doubletFinder loom
+#' @examples doubletFinder(pbmc_loom, expected.doublets = 1000, proportion.artificial = 0.25,
+#'               proportion.NN = 0.01, gene.names = "row_attrs/Accession",
+#'               cell.names = "col_attrs/CellID",
+#'               overwrite = TRUE, pcs.compute = 10, online.pca = FALSE)
+
+doubletFinder.loom <- function(obj, expected.doublets = 0,
+                               proportion.artificial = 0.25, proportion.NN = 0.01,
+                               gene.names = "row_attrs/gene_names",
+                               cell.names = "col_attrs/cell_names",
+                               overwrite = FALSE, pcs.compute = 5,
+                               chunk.size = NULL, chunk.dims = NULL, ...) {
+  if (expected.doublets == 0) {  stop("Need to set number of expected doublets...")  }
+  ## Step 1: Generate artificial doublets from loom object input
+  print("Creating artificial doublets...")
+  real.cells <- obj[[cell.names]][]
+  genes <- obj[[gene.names]][]
+  n_real.cells <- length(real.cells)
+  n_doublets <- round(n_real.cells/(1 - proportion.artificial) - n_real.cells)
+  nCells <- n_real.cells + n_doublets
+  cells_both <- c(real.cells, paste0("X", 1:n_doublets))
+
+  # Make temporary loom file for data with simulated doublets
+  # Generate doublets
+  real_ind1 <- sample(1:n_real.cells, n_doublets, replace = TRUE)
+  real_ind2 <- sample(1:n_real.cells, n_doublets, replace = TRUE)
+  # If not chunked, this will load about half of the data into memory
+  # To conserve memory, get doublets by chunk
+  # As a rule of thumb, do calculation for 50 cells at a time
+  doublet_mat <- matrix(0, n_doublets, length(genes))
+  if (n_doublets <= 50) {
+    doublet_mat <-
+      (obj[["matrix"]][real_ind1,] + obj[["matrix"]][real_ind2,]) / 2
+  } else {
+    n_chunks <- floor(n_doublets / 50) # the last chunk is added separately
+    pb <- txtProgressBar(style = 3)
+    for (i in 1:n_chunks) {
+      ind1 <- 1 + (i - 1) * 50
+      ind2 <- i * 50
+      doublet_mat[ind1:ind2,] <-
+        (obj[["matrix"]][real_ind1[ind1:ind2],] + obj[["matrix"]][real_ind2[ind1:ind2],]) / 2
+      setTxtProgressBar(pb, i / n_chunks)
+    }
+    close(pb)
+    remainder <- n_chunks %% 50
+    if (remainder > 0) {
+      doublet_mat[(n_chunks * 50 + 1):n_doublets,] <-
+        (obj[["matrix"]][real_ind1[(ind2 + 1):n_doublets],] + obj[["matrix"]][real_ind2[(ind2 + 1):n_doublets],]) / 2
+    }
+  }
+  print("Writing artificial doublets to loom file")
+  doublets <- create_doublet_loom(n_doublets, doublet_mat, obj, overwrite = overwrite)
```

Comment:
As a temporary workaround to the NormalizeData issue, have you considered first generating a doublet-only Seurat object? Doublet-only Seurat objects would rarely be >20% of the original loom object (likely only when multiplexing methods are used) so may not be as memory-intensive. The original scale factor could be used, and then the doublet seurat object could be converted to loom format and then merged?

Reply:
I profiled memory usage. The main problem is not with generating doublets, but with hdf5, probably due to keeping cache of chunks and there are many chunks. The functions that use memory without returning are all C functions. I can try a different implementation, using the `add.cell` method of loom objects rather than `combine`, to minimize chunked operations before the real stuff, like `NormalizeData`. The problem is that `loomR` is still in development, and `add.cell` doesn't play nicely with 2D column attributes. I'll need to fix that in order to use `add.cell`.

Reply:
I did generate a doublet only loom file first, and then combined it with the loom file that has real cells. I have to combine them in order for the PCA cell embeddings to make sense and compare real cells to artificial doublets. The main memory problem arose from that combining step.